extensions/podman/scripts/download.ts

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env node
+/**********************************************************************
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+import { https } from 'follow-redirects';
+import * as fs from 'node:fs';
+import * as path from 'node:path';
+
+const tools = require('../src/podman.json');
+
+import * as hasha from 'hasha';
+
+const platform = process.platform;
+
+const MAX_DOWNLOAD_ATTEMPT = 3;
+let downloadAttempt = 0;
+
+// to make this file a module
+export {};
+
+function downloadFile(url, dest): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const file = fs.createWriteStream(dest, { flags: 'wx' });
+
+    const request = https.get(url, { headers: { 'User-Agent': 'Podman Desktop' } }, response => {
+      if (response.statusCode === 200) {
+        response.pipe(file);
+      } else {
+        file.close();
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        fs.unlink(dest, () => {}); // Delete temp file
+        reject(`Server responded with ${response.statusCode}: ${response.statusMessage}`);
+      }
+    });
+
+    request.on('error', err => {
+      file.close();
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      fs.unlink(dest, () => {}); // Delete temp file
+      reject(err.message);
+    });
+
+    file.on('finish', () => {
+      resolve();
+    });
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    file.on('error', err => {
+      file.close();
+      console.error(err);
+      if ((err as unknown as { code: string }).code === 'EEXIST') {
+        reject('File already exists');
+      } else {
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        fs.unlink(dest, () => {}); // Delete temp file
+        reject(err.message);
+      }
+    });
+  });
+}
+async function downloadPackage(url, fileName): Promise<string | undefined> {
+  if (url && fileName) {
+    const destDir = path.resolve(__dirname, '..', 'assets');
+    if (!fs.existsSync(destDir)) {
+      fs.mkdirSync(destDir);
+    }
+    const destFile = path.resolve(destDir, fileName);
+    if (!fs.existsSync(destFile)) {
+      console.log(`Downloading Podman package from ${url}`);
```

Comment:
Hello, how does it work for developers using watch mode, etc
if the folder is removed at every launch probably we should cache the files to avoid to fetch them multiple times in development mode.

Reply:
It is cached in `assets` directory, we download msi once